runtime/dart_isolate.cc

```diff
@@ -280,8 +283,136 @@ std::weak_ptr<DartIsolate> DartIsolate::CreateRootIsolate(
+                  Dart_IsolateGroupData(parent_isolate))));
+
+  Settings settings((*isolate_group_data)->GetSettings());
+
+  // PlatformIsolate.spawn should behave like Isolate.spawn when unhandled
+  // exceptions happen (log the exception, but don't terminate the app). But the
+  // default unhandled_exception_callback may terminate the app, because it is
+  // only called for the root isolate (child isolates are managed by the VM and
+  // have a different error code path). So override it to simply log the error.
+  settings.unhandled_exception_callback = [errors_fatal](
+                                              const std::string& error,
+                                              const std::string& stack_trace) {
+    FML_LOG(ERROR) << "Unhandled exception:\n" << error << "\n" << stack_trace;
+    if (errors_fatal) {
+      Dart_ShutdownIsolate();
+    }
+    return true;
+  };
+
+  UIDartState::Context context(task_runners);
+  context.advisory_script_uri = (*isolate_group_data)->GetAdvisoryScriptURI();
+  context.advisory_script_entrypoint =
+      (*isolate_group_data)->GetAdvisoryScriptEntrypoint();
+  auto isolate_data = std::make_unique<std::shared_ptr<DartIsolate>>(
+      std::shared_ptr<DartIsolate>(new DartIsolate(settings,  // settings
+                                                   false,     // is_root_isolate
+                                                   true,  // is_platform_isolate
+                                                   context)));  // context
+  (*isolate_data)->platform_isolate_manager_ = platform_isolate_manager;
+
+  IsolateMaker isolate_maker =
+      [parent_isolate, debug_name](
+          std::shared_ptr<DartIsolateGroupData>* isolate_group_data,
+          std::shared_ptr<DartIsolate>* isolate_data, Dart_IsolateFlags* flags,
+          char** error) {
+        return Dart_CreateIsolateInGroup(
+            /*group_member=*/parent_isolate,
+            /*name=*/debug_name,
+            /*shutdown_callback=*/
+            reinterpret_cast<Dart_IsolateShutdownCallback>(
+                DartIsolate::SpawnIsolateShutdownCallback),
+            /*cleanup_callback=*/
+            reinterpret_cast<Dart_IsolateCleanupCallback>(
+                DartIsolateCleanupCallback),
+            /*child_isolate_data=*/isolate_data,
+            /*error=*/error);
+      };
+  Dart_Isolate platform_isolate = CreateDartIsolateGroup(
+      std::move(isolate_group_data), std::move(isolate_data), nullptr, error,
+      isolate_maker);
+
+  Dart_EnterIsolate(parent_isolate);
+
+  if (*error) {
+    return nullptr;
+  }
+
+  if (!platform_isolate_manager->RegisterPlatformIsolate(platform_isolate)) {
+    // The PlatformIsolateManager was shutdown while we were creating the
+    // isolate. This means that we're shutting down the engine. Do nothing. The
+    // ordinary engine shut down procedure will clean up the isolate.
+    FML_LOG(ERROR) << "Shutdown during platform isolate creation";
+    return nullptr;
```

Comment:
Does the `platform_isolate` need to be destroyed here?

Reply:
I thought the ordinary Dart VM shutdown procedure also shutdown all the isolates. Is that not the case? Does the flutter engine manually shut down the isolates?

Reply:
The engine shuts down isolates explicitly (for example, the call to `root_isolate->Shutdown()` in the `RuntimeController` destructor).

Multiple Flutter engine `Shell` instances can exist in the same process.  So deleting a `Shell` will not terminate the Dart VM.

Given that, is there a risk that the `platform_isolate` will be leaked if `DartIsolate::CreatePlatformIsolate` fails to register it with the `PlatformIsolateManager`?

Reply:
Yep, as it's written here this will leak the platform isolate in this case. I've changed this branch to explicitly shutdown the platform isolate. I don't think any of the other error branches leak the isolate.